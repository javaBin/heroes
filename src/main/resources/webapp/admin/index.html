<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="UTF-8">
        <title>javaBin Heroes Admin</title>
        <link rel="stylesheet" href="css/heroes-admin.css"/>
        <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.5/css/bootstrap.min.css"
              integrity="sha512-dTfge/zgoMYpP7QbHy4gWMEGsbsdZeCXz7irItjcC3sPUFtf0kuFbDz/ixG7ArTxmDjLXDmezHubeNikyKGVyQ=="
              crossorigin="anonymous"/>
    </head>
    <body>
        <div id="heroesadmin"></div>
        <script src="../node_modules/react/dist/react.js"></script>
        <script src="../node_modules/react-dom/dist/react-dom.js"></script>
        <script src="../node_modules/react-bootstrap/dist/react-bootstrap.js"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/babel-core/5.8.23/browser.min.js"></script>

        <script type="text/babel" src="js/app.js"></script>
    </body>
</html>