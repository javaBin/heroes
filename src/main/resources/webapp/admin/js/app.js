"use strict";

var heroes = [
  {navn: "Jaran Flaath", bidrag: "Nestleder", email: "j.f@g.com", fraDato: "2015-01-01"},
  {navn: "Anders Karlsen", bidrag: "Aktiv", email: "a.k@g.com", fraDato: "2015-01-01"},
  {navn: "Bjørn Hamre", bidrag: "Styremedlem", email: "b.h@g.com", fraDato: "2015-01-01"}
];

var Button = ReactBootstrap.Button;
var Input = ReactBootstrap.Input;
var Grid = ReactBootstrap.Grid;
var Row = ReactBootstrap.Row;
var Col = ReactBootstrap.Col;
var Table = ReactBootstrap.Table;

var NewHeroForm = React.createClass({
  render: function () {
    return (
      <div className="heroes-admin-add">
        <Grid>
          <Row>
            <Col md={6}>
              <Input type="text" label="Navn" placeholder="Heltens navn"/>
            </Col>
            <Col md={6}>
              <Input type="email" label="Email" placeholder="hero@java.no"/>
            </Col>
          </Row>
          <Row>
            <Col md={6}>
              <Input type="select" label="Heltetype">
                <option value="styremedlem">Styremedlem</option>
                <option value="foredragsholder-jz">Foredragsholder på JavaZone</option>
                <option value="foredragsholder">Foredragsholder på javaBin</option>
                <option value="regionsleder">Regionsleder</option>
                <option value="aktiv">Aktiv</option>
              </Input>
            </Col>
            <Col md={6}>
              <Input type="date" label="Fra dato"/>
            </Col>
          </Row>
        </Grid>
        <Button bsStyle="info">Legg til</Button>
      </div>
    );
  }
});


var HeroesList = React.createClass({
  render: function () {

    var heroComponents = this.props.data.map(function (hero) {
      return (
        <tr>
          <td>
            <img height="25" src="images/hero.png"/>
          </td>
          <td>
            {hero.navn}
          </td>
          <td>
            {hero.bidrag}
          </td>
        </tr>
      );
    });
    return (
      <div className="heroes-list">
        <h2>Helter</h2>
        <Table striped bordered condensed hover>
          <thead>
            <tr>
              <th></th>
              <th>Navn</th>
              <th>Heltetype</th>
            </tr>
          </thead>
          <tbody>
            {heroComponents}
          </tbody>
        </Table>
      </div>
    );
  }
});

var HeroesAdmin = React.createClass({
  render: function () {
    return (
      <div className="heroes-admin-container">
        <Grid>
          <Row>
            <Col md={12}>
              <h1>javaBin Heroes</h1>
              <NewHeroForm />
              <HeroesList data={this.props.data}/>
            </Col>
          </Row>
        </Grid>
      </div>
    );

  }
});
ReactDOM.render(<HeroesAdmin data={heroes}/>, document.getElementById("heroesadmin"));
