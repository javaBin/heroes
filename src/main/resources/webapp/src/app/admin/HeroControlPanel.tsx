import React, { ChangeEvent, FormEvent, MouseEvent } from "react";
import {
    Achievement, achievementName, allAchievements, Hero, HeroAchievement, Person,
} from "../../services/heroService";

export class HeroControlPanel extends React.Component<{
    heroes: Hero[], people: Person[], prefix: string,
}, {
    selectedHero?: Hero, addHero?: boolean, action?: string,
    heroes: Hero[], actionTargetId?: string,
}> {

    constructor(props: {heroes: Hero[], people: Person[], prefix: string}) {
        super(props);
        this.state = {heroes: props.heroes};
    }

    async componentDidMount() {
        window.addEventListener("hashchange", this.handleHashChange);
        this.setHash(window.location.hash);
    }

    componentWillUnmount() {
        window.removeEventListener("hashchange", this.handleHashChange);
    }

    handleHashChange = (event: HashChangeEvent) => {
        this.setHash(window.location.hash);
    }

    setHash = async (hash: string) => {
        const [prefix, controller, idOrAction, subAction, actionTargetId] = hash.split("/");
        if (idOrAction === "add") {
            await this.setState({addHero: true});
        } else if (controller === "heroes" && idOrAction) {
            const selectedHero = this.state.heroes.find(p => p.id === idOrAction);
            if (!selectedHero) {
                window.location.hash = prefix + "/heroes";
            } else {
                await this.setState({action: subAction, selectedHero, actionTargetId});
            }
        } else {
            await this.setState({selectedHero: undefined, addHero: false, actionTargetId: undefined});
        }
    }

    handleAddHero = (hero: Hero) => {
        hero.id = "" +  (this.state.heroes.length + 1);
        this.setState({heroes: this.state.heroes.concat(hero)});
        window.location.hash = this.props.prefix + "/heroes";
    }

    handleUpdateHero = (heroId: string, update: Partial<Hero>) => {
        const index = this.state.heroes.findIndex(h => h.id === heroId);
        if (index !== undefined) {
            const hero = {...this.state.heroes[index], ...update};
            this.state.heroes[index] = hero;
            this.setState({heroes: this.state.heroes});
        }
        window.location.hash = this.props.prefix + "/heroes/" + heroId;
    }

    handleAddAchievement = (heroId: string, achievement: HeroAchievement) => {
        const index = this.state.heroes.findIndex(h => h.id === heroId);
        if (index !== undefined) {
            achievement.id = "" + (this.state.heroes[index].achievements.length + 1);
            this.state.heroes[index].achievements.push(achievement);
            this.setState({heroes: this.state.heroes});
        }
        window.location.hash = this.props.prefix + "/heroes/" + heroId;
    }

    handleUpdateAchievement = (heroId: string, achievementId: string, achievement: any) => {
        const index = this.state.heroes.findIndex(h => h.id === heroId);
        if (index !== undefined) {
            const hero = this.state.heroes[index];
            const achievementIndex = hero.achievements.findIndex(a => a.id === achievementId);
            hero.achievements[achievementIndex] = {...hero.achievements[achievementIndex], ...achievement};
            this.setState({heroes: this.state.heroes});
        }
        window.location.hash = this.props.prefix + "/heroes/" + heroId;
    }

    handleDeleteAchievement = (heroId: string, achievementId: string) => {
        const index = this.state.heroes.findIndex(h => h.id === heroId);
        if (index !== undefined) {
            const hero = this.state.heroes[index];
            const achievementIndex = hero.achievements.findIndex(a => a.id === achievementId);
            hero.achievements.splice(achievementIndex);
            this.setState({heroes: this.state.heroes});
        }
        window.location.hash = this.props.prefix + "/heroes/" + heroId;
    }

    handleCancelAddHero = () => {
        window.location.hash = this.props.prefix + "/heroes";
    }

    render() {
        if (this.state.addHero) {
            return <AddHeroView
                people={this.props.people}
                onSubmit={this.handleAddHero}
                onCancel={this.handleCancelAddHero}
            />;
        }  else if (!this.state.selectedHero) {
            return <HeroListView heroes={this.state.heroes} prefix={this.props.prefix} />;
        } else {
            return <HeroView
                hero={this.state.selectedHero}
                action={this.state.action}
                actionTargetId={this.state.actionTargetId}
                prefix={this.props.prefix}
                onSubmit={this.handleUpdateHero}
                onAddAchievement={this.handleAddAchievement}
                onUpdateAchievement={this.handleUpdateAchievement}
                onDeleteAchievement={this.handleDeleteAchievement}
            />;
        }
    }
}

export class AddHeroView extends React.Component<{
    people: Person[],
    onSubmit: (hero: Hero) => void,
    onCancel: () => void,
}, Partial<Person>> {
    constructor(props: {people: Person[], onSubmit: (hero: Hero) => void, onCancel: () => void}) {
        super(props);
        this.state = {};
    }

    handleSelectSlackPerson = (e: ChangeEvent<HTMLSelectElement>) => {
        const {value} = e.target;
        const person = this.props.people.find(p => p.email === value);
        if (person) {
            this.setState({...person});
        }
    }

    handleCancel = (e: MouseEvent) => {
        e.preventDefault();
        this.props.onCancel();
    }

    handleSubmit = (e: FormEvent) => {
        e.preventDefault();
        const {name, email, twitter} = this.state;
        if (email && name) {
            this.props.onSubmit({name, email, twitter, achievements: [], published: false});
        }
    }

    render() {
        return <form onSubmit={this.handleSubmit}>
            <h2>Add a hero</h2>

            <div>
                <label>
                    Select from slack:
                    <select onChange={this.handleSelectSlackPerson} value={this.state.email} autoFocus>
                        <option></option>
                        {this.props.people.map(p =>
                            <option key={p.email} value={p.email}>{p.name} &lt;{p.email}&gt;</option>,
                        )}
                    </select>
                </label>
            </div>
            <div>
                <label>
                    Name:
                    <input value={this.state.name} onChange={e => this.setState({name: e.target.value})} />
                </label>
            </div>
            <div>
                <label>
                    Email:
                    <input value={this.state.email} onChange={e => this.setState({email: e.target.value})} />
                </label>
            </div>
            <div>
                <label>
                    Twitter:
                    <input value={this.state.twitter} onChange={e => this.setState({twitter: e.target.value})} />
                </label>
            </div>
            <button>Submit</button>
            <button onClick={this.handleCancel}>Cancel</button>
        </form>;
    }
}

interface HeroEditProps {
    hero: Hero;
    action?: string;
    actionTargetId?: string;
    prefix: string;
    onSubmit: (id: string, hero: Partial<Hero>) => void;
    onAddAchievement: (heroId: string, achievement: any) => void;
    onUpdateAchievement: (heroId: string, achievementId: string, achievement: any) => void;

    onDeleteAchievement: (heroId: string, achievementId: string) => void;
}

export class HeroAchievementList extends React.Component<{
    hero: Hero, achievements: any[], prefix: string,
    onDeleteAchievement: (heroId: string, achievementId: string) => void,
}> {

    handleDeleteAchievement = (e: MouseEvent, achievementId: string) => {
        e.preventDefault();
        this.props.onDeleteAchievement(this.props.hero.id!, achievementId);
    }

    renderAchievement = (achievement: HeroAchievement) => {
        return <li key={achievement.label}>
            {achievement.label} [
                <a href={this.props.prefix + "/heroes/" + this.props.hero.id + "/achievement/" + achievement.id}>
                Edit
                </a>]
                [
                <a
                    href="#"
                    onClick={e => this.handleDeleteAchievement(e, achievement.id!)}
                    className="deleteAchievementLink"
                >
                Delete
                </a>]
        </li>;
    }

    render() {
        return <>
            <h3>Achievements</h3>

            <ul>
                {this.props.achievements.map(this.renderAchievement)}
            </ul>
        </>;
    }
}

export class HeroView extends React.Component<HeroEditProps, Partial<Hero>> {
    constructor(props: HeroEditProps) {
        super(props);
        const {hero} = props;
        this.state = {
            achievements: hero.achievements,
            email: hero.email,
            name: hero.name,
            published: hero.published,
            twitter: hero.twitter,
        };
    }

    handleSubmit = (e: FormEvent) => {
        e.preventDefault();
        const {email, name, twitter} = this.state;
        const hero = {email, name, twitter};
        this.props.onSubmit(this.props.hero.id!, hero);
    }

    render() {
        const {hero, prefix, onDeleteAchievement} = this.props;
        const {name, email, twitter, achievements} = this.state;
        return <>
            <h3><a href={prefix}>Back</a></h3>
            <h2>{hero.name}</h2>

            {!this.props.action && <>
                <ul>
                    <li>Email: {hero.email}</li>
                    <li>Twitter: {hero.twitter}</li>
                    <li><a href={prefix + "/heroes/" + hero.id + "/edit"}>Update</a></li>
                </ul>

                {achievements && <HeroAchievementList
                    hero={hero}
                    achievements={achievements}
                    prefix={prefix}
                    onDeleteAchievement={onDeleteAchievement}
                />}

                <p>
                    <a href={prefix + "/heroes/" + hero.id + "/addAchievement"}>Add achievement</a>
                </p>
            </>}

            {this.props.action === "edit" && <>
                <form onSubmit={this.handleSubmit}>
                {hero.avatar && <img src={hero.avatar} alt={"Picture of " + hero.name} />}
                    <h3>Information</h3>
                    <ul>
                        <li>Display name:
                            <input
                                autoFocus
                                value={name}
                                onChange={e => this.setState({name: e.target.value})}
                            />
                        </li>
                        <li>Email address:
                            <input value={email} onChange={e => this.setState({email: e.target.value})} />
                        </li>
                        <li>Twitter handle:
                            <input
                                value={twitter}
                                onChange={e => this.setState({twitter: e.target.value})}
                            />
                        </li>
                    </ul>

                    <button>Lagre</button>
                    <a href={prefix + "/heroes/" + hero.id}>Back</a>
                </form>
            </>}

            {this.props.action === "addAchievement" &&
                <AddHeroAchievement
                    hero={hero}
                    onSubmit={this.props.onAddAchievement}
                    prefix={prefix}
                />
            }
            {this.props.action === "achievement" && this.props.actionTargetId !== undefined &&
                <HeroAchievementEditView
                    hero={hero}
                    achievementId={this.props.actionTargetId}
                    onSubmit={this.props.onUpdateAchievement}
                    prefix={prefix}
                />
            }
        </>;
    }
}

interface HeroAchievementEditProps {
    hero: Hero;
    achievementId: string;
    prefix: string ;
    onSubmit: (heroId: string, achievementId: string, o: any) => void;
}

export class HeroAchievementEditView extends React.Component<HeroAchievementEditProps> {
    handleSave = (update: any) => {
        this.props.onSubmit(this.props.hero.id!, this.props.achievementId, update);
    }

    render() {
        const {hero, achievementId} = this.props;
        const achievement = hero.achievements.find(a => a.id === achievementId)!;
        const DetailView = achievementDetail(achievement.type);
        return <form>
            <DetailView hero={hero} onSave={this.handleSave} achievement={achievement} />
        </form>;
    }
}

interface HeroAchievementProps {
    hero: Hero;
    achievement?: any;
    onSave(o: any): void;
}

export class JavaZoneSpeakerAchievementDetails extends React.Component<
    HeroAchievementProps & {achievement?: {year?: string, title?: string}}, {
    year?: string, title: string,
}> {
    years: string[];
    constructor(props: HeroAchievementProps & {achievement?: {year?: string, title?: string}}) {
        super(props);
        this.years = [
            "2018", "2017", "2016", "2015", "2014",
            "2013", "2012", "2011", "2010", "2009",
            "2008",
        ];
        this.state = { year: this.years[0], title: "", ...props.achievement };
    }

    handleSubmit = (e: FormEvent) => {
        const {year, title} = this.state;
        const label = "Speaker at JavaZone " + year + ": " + title;
        this.props.onSave({ year, title, label });
        e.preventDefault();
    }

    render() {
        return <div>
            <h4>JavaZone foredragsholder</h4>
            <label>
                JavaZone year
                <select value={this.state.year} onChange={e => this.setState({year: e.target.value})} autoFocus>
                    {this.years.map(y => <option value={y} key={y}>{y}</option>)}
                </select>
            </label>
            <label>
                Title:
                <input value={this.state.title} onChange={e => this.setState({title: e.target.value})} />
            </label>
            <button onClick={this.handleSubmit} disabled={!this.state.title.length}>Submit</button>
        </div>;
    }
}

export class JavaBinSpeakerAchievementDetails extends React.Component<
    HeroAchievementProps & {achievement?: {date?: string, title?: string}}, {
    date?: string,
    title: string,
}> {
    constructor(props: HeroAchievementProps & {achievement?: {date?: string, title?: string}}) {
        super(props);
        this.state = {title: "", ...props.achievement};
    }
    handleSubmit = (e: FormEvent) => {
        const {date, title} = this.state;
        const label = `JavaBin speaker ${date}: ${title}`;
        this.props.onSave({ date, title, label });
        e.preventDefault();
    }

    render() {
        return <>
            <h3>JavaBin usergroup speaker</h3>
            <label>
                Title:
                <input value={this.state.title} onChange={e => this.setState({title: e.target.value})} />
            </label>
            <label>
                Date:
                <input type="date" value={this.state.date} onChange={e => this.setState({date: e.target.value})} />
            </label>
            <button onClick={this.handleSubmit} disabled={!this.state.title.length || !this.state.date}>Submit</button>
        </>;
    }
}

export class BoardMemberAchievementDetails extends React.Component<
    HeroAchievementProps & {achievement?: {year?: string, role?: string}}, {
    year?: string, role: string,
}> {
    years: string[];
    roles: string[];
    constructor(props: HeroAchievementProps & {achievement?: {year?: string, role?: string}}) {
        super(props);
        this.years = [
            "2018", "2017", "2016", "2015", "2014",
            "2013", "2012", "2011", "2010", "2009",
            "2008",
        ];
        this.roles = [
            "board member", "vice chair", "chair",
        ];
        this.state = { role: this.roles[0], year: this.years[0], ...props.achievement };
    }

    handleSubmit = (e: FormEvent) => {
        const {year, role} = this.state;
        const label = role + " i styret " + year;
        this.props.onSave({ year, role, label });
        e.preventDefault();
    }

    render() {
        return <>
            <h3>JavaBin board member</h3>
            <label>
                Role
                <select value={this.state.role} onChange={e => this.setState({role: e.target.value})}>
                    {this.roles.map(r => <option value={r} key={r}>{r}</option>)}
                </select>
            </label>
            <label>
                Elected year
                <select value={this.state.year} onChange={e => this.setState({year: e.target.value})}>
                    {this.years.map(y => <option value={y} key={y}>{y}</option>)}
                </select>
            </label>
            <button onClick={this.handleSubmit} disabled={!this.state.year || !this.state.role.length}>Submit</button>
        </>;
    }
}

class EmptyAchievementDetails extends React.Component<HeroAchievementProps> {
    render() {
        return <div></div>;
    }
}

function achievementDetail(achievementType?: Achievement): React.ComponentType<HeroAchievementProps> {
    switch (achievementType) {
    case Achievement.foredragsholder_javabin:   return JavaBinSpeakerAchievementDetails;
    case Achievement.foredragsholder_jz:        return JavaZoneSpeakerAchievementDetails;
    case Achievement.styre:                     return BoardMemberAchievementDetails;
    case undefined:                             return EmptyAchievementDetails;
    }
}

export class AddHeroAchievement extends React.Component<{
    hero: Hero, prefix: string, onSubmit(heroId: string, o: any): void,
}, {
    achievementType?: Achievement, achievementTypeString?: string,
}> {
    constructor(props: {
        hero: Hero, prefix: string, onSubmit: (heroId: string, o: any) => void, achievementTypes: Achievement[],
    }) {
        super(props);
        this.state = {};
    }

    renderAchievementType = (achivementType: Achievement) => {
        return <option
            key={achivementType}
            value={Achievement[achivementType]}
        >{achievementName(achivementType)}</option>;
    }

    handleSave = (o: any) => {
        this.props.onSubmit(this.props.hero.id!, {type: this.state.achievementType, ...o});
        this.setState({achievementType: undefined});
    }

    handleChangeAchievementType = (e: ChangeEvent<HTMLSelectElement>) => {
        const {value} = e.target;
        const achievementType: Achievement = (Achievement as any)[value];
        this.setState({achievementType, achievementTypeString: value});
    }

    render() {
        const {achievementTypeString} = this.state;
        const {hero} = this.props;

        const DetailComponent = achievementDetail(this.state.achievementType);

        return <>
            <h3>Please Add achievement</h3>
            <form>
                <label>
                    Achievement:
                    <select autoFocus={true} value={achievementTypeString} onChange={this.handleChangeAchievementType}>
                        <option></option>
                        {allAchievements().map(this.renderAchievementType)}
                    </select>
                </label>
                <DetailComponent hero={hero} onSave={this.handleSave} ></DetailComponent>
                <a href={this.props.prefix + "/heroes/" + hero.id}>Back</a>
            </form>
        </>;
    }
}

export class HeroListView extends React.Component<{heroes: Hero[], prefix: string}> {

    renderHero = (hero: Hero) => {
        return <li key={hero.id}><a href={this.props.prefix + "/heroes/" + hero.id}>{hero.name}</a></li>;
    }

    render() {
        return <>
            <h2>Here are all the heroes</h2>
            <ul>
                {this.props.heroes.map(this.renderHero)}
            </ul>

            <a href={this.props.prefix + "/heroes/add"}>Add a hero</a>
        </>;
    }
}
