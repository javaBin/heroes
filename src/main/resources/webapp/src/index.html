<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <title>Document</title>
    <link rel="stylesheet" href="css/heroes-admin.css"/>
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.5/css/bootstrap.min.css"
        integrity="sha512-dTfge/zgoMYpP7QbHy4gWMEGsbsdZeCXz7irItjcC3sPUFtf0kuFbDz/ixG7ArTxmDjLXDmezHubeNikyKGVyQ=="
        crossorigin="anonymous"/>
</head>
<body>
    <div id="app">Loading...</div>
    <script src="index.tsx"></script>
    
</body>
</html>
