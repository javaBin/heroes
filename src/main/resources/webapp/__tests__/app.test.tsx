import React from "react";
import renderer, { ReactTestInstance, ReactTestRenderer } from "react-test-renderer";
import { App, HeroList } from "../src/app";
import { HeroControlPanel } from "../src/app/admin/HeroControlPanel";
import { MockHeroService } from "../src/services/mockHeroService";

function promiseCompletion() {
    return new Promise((resolve) => setImmediate(resolve));
}

const mockHeroService = new MockHeroService();
jest.mock("../src/services/heroServiceHttp", () => ({
    HeroServiceHttp: () => mockHeroService,
}));

// tslint:disable-next-line:no-eval
eval(`global["window"] = {
    location: {
        hash: ""
    },
    addEventListener() {},
    removeEventListener() {}
}`);

function findComponent(ancestor: ReactTestRenderer, type: React.ReactType): ReactTestInstance {
    expect(ancestor.root.findAll(t => true).map(t => t.type)).toContain(type);
    return ancestor.root.findByType(type);

}

describe("app", () => {
    it("renders hero list", async () => {
        const heroes = [
            {name: "Alice", email: "alice@example.com", achievements: [], achievement: "", published: true, id: "1"},
            {name: "Bob", email: "bob@example.com", achievements: [], achievement: "", published: true, id: "2"},
        ];
        mockHeroService.fetchPublicHeroes = async () => heroes;
        const app = renderer.create(<App heroService={mockHeroService} />);
        await promiseCompletion();
        expect(findComponent(app, HeroList).props).toHaveProperty("heroes", heroes);
        expect(app.toJSON()).toMatchSnapshot();
    });

    it("renders admin page", async () => {
        window.location.hash = "#admin";
        const app = renderer.create(<App heroService={mockHeroService} />);
        await promiseCompletion();

        expect(findComponent(app, HeroControlPanel).props).toHaveProperty("heroService", mockHeroService);
        expect(app.toJSON()).toMatchSnapshot();
    });
});
