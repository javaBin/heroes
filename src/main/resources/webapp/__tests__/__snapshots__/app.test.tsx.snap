// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app renders admin page 1`] = `
Array [
  <header
    className="MuiPaper-root-141 MuiPaper-elevation4-147 MuiAppBar-root-132 MuiAppBar-positionStatic-136 MuiAppBar-colorPrimary-139"
  >
    <div
      className="MuiToolbar-root-168 MuiToolbar-regular-170 MuiToolbar-gutters-169"
    >
      <a
        className="MuiButtonBase-root-178 MuiIconButton-root-172 MuiIconButton-colorInherit-173"
        href="#"
        onBlur={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex="0"
      >
        <span
          className="MuiIconButton-label-177"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-181"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
            />
          </svg>
        </span>
        <span
          className="MuiTouchRipple-root-226"
        />
      </a>
      <h6
        className="MuiTypography-root-190 MuiTypography-h6-207 MuiTypography-colorInherit-219"
        style={
          Object {
            "flexGrow": 1,
          }
        }
      >
        Heroes
      </h6>
      <a
        className="MuiButtonBase-root-178 MuiIconButton-root-172 MuiIconButton-colorInherit-173"
        href="#profile"
        onBlur={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex="0"
      >
        <span
          className="MuiIconButton-label-177"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-181"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
            />
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
          </svg>
        </span>
        <span
          className="MuiTouchRipple-root-226"
        />
      </a>
    </div>
  </header>,
  <div
    className="MuiPaper-root-141 MuiPaper-elevation2-145 MuiPaper-rounded-142 class_1-paper-234"
  >
    <h4
      className="MuiTypography-root-190 MuiTypography-h4-205"
    >
      Heroes
    </h4>
    <ul
      className="MuiList-root-235 MuiList-padding-236"
    >
      <div
        className="MuiButtonBase-root-178 MuiListItem-root-239 MuiListItem-default-242 MuiListItem-gutters-247 MuiListItem-button-248"
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex="0"
      >
        <div
          className="MuiListItemIcon-root-251"
        >
          <div
            className="MuiAvatar-root-252 MuiAvatar-colorDefault-253"
          />
        </div>
        <div
          className="MuiListItemText-root-255"
        >
          <span
            className="MuiTypography-root-190 MuiTypography-body1-199 MuiListItemText-primary-258"
          >
            Johannes
          </span>
          <p
            className="MuiTypography-root-190 MuiTypography-body2-198 MuiTypography-colorTextSecondary-223 MuiListItemText-secondary-259"
          >
            johannes@brodwall.com
          </p>
        </div>
        <span
          className="MuiTouchRipple-root-226"
        />
      </div>
    </ul>
    <a
      className="MuiButtonBase-root-178 MuiButton-root-261 MuiButton-contained-272 MuiButton-containedPrimary-273 MuiButton-raised-275 MuiButton-raisedPrimary-276"
      href="#admin/heroes/add"
      onBlur={[Function]}
      onContextMenu={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex="0"
    >
      <span
        className="MuiButton-label-262"
      >
        Add hero
      </span>
      <span
        className="MuiTouchRipple-root-226"
      />
    </a>
  </div>,
]
`;

exports[`app renders hero list 1`] = `
Array [
  <header
    className="MuiPaper-root-10 MuiPaper-elevation4-16 MuiAppBar-root-1 MuiAppBar-positionStatic-5 MuiAppBar-colorPrimary-8"
  >
    <div
      className="MuiToolbar-root-37 MuiToolbar-regular-39 MuiToolbar-gutters-38"
    >
      <a
        className="MuiButtonBase-root-47 MuiIconButton-root-41 MuiIconButton-colorInherit-42"
        href="#"
        onBlur={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex="0"
      >
        <span
          className="MuiIconButton-label-46"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-50"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
            />
          </svg>
        </span>
        <span
          className="MuiTouchRipple-root-97"
        />
      </a>
      <h6
        className="MuiTypography-root-59 MuiTypography-h6-76 MuiTypography-colorInherit-88"
        style={
          Object {
            "flexGrow": 1,
          }
        }
      >
        Heroes
      </h6>
      <a
        className="MuiButtonBase-root-47 MuiIconButton-root-41 MuiIconButton-colorInherit-42"
        href="#profile"
        onBlur={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex="0"
      >
        <span
          className="MuiIconButton-label-46"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-50"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
            />
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
          </svg>
        </span>
        <span
          className="MuiTouchRipple-root-97"
        />
      </a>
    </div>
  </header>,
  <div>
    <div
      className="MuiPaper-root-10 MuiPaper-elevation2-14 MuiPaper-rounded-11 class_1-paper-96"
    >
      <h4
        className="MuiTypography-root-59 MuiTypography-h4-74"
      >
        Heroes
      </h4>
      <div
        className="heroes-list"
      >
        <ul
          className="MuiList-root-104 MuiList-padding-105"
        >
          <li
            className="MuiListItem-root-108 MuiListItem-default-111 MuiListItem-dense-112 MuiListItem-gutters-116"
            disabled={false}
          >
            <div
              className="MuiAvatar-root-123 MuiAvatar-colorDefault-124"
            />
            <div
              className="MuiListItemText-root-126"
            >
              <span
                className="MuiTypography-root-59 MuiTypography-body1-68 MuiListItemText-primary-129"
              >
                Alice
              </span>
              <p
                className="MuiTypography-root-59 MuiTypography-body2-67 MuiTypography-colorTextSecondary-92 MuiListItemText-secondary-130"
              >
                
              </p>
            </div>
          </li>
          <li
            className="MuiListItem-root-108 MuiListItem-default-111 MuiListItem-dense-112 MuiListItem-gutters-116"
            disabled={false}
          >
            <div
              className="MuiAvatar-root-123 MuiAvatar-colorDefault-124"
            />
            <div
              className="MuiListItemText-root-126"
            >
              <span
                className="MuiTypography-root-59 MuiTypography-body1-68 MuiListItemText-primary-129"
              >
                Bob
              </span>
              <p
                className="MuiTypography-root-59 MuiTypography-body2-67 MuiTypography-colorTextSecondary-92 MuiListItemText-secondary-130"
              >
                
              </p>
            </div>
          </li>
        </ul>
      </div>
    </div>
  </div>,
]
`;
